### ISSUE TITLE
FluxEngine

### ISSUE THREAD

**Issue by ChickenMan540** [davidgiven/fluxengine#67]
Hi David,
I'm trying to program my CY8CKIT-059.  I'm running Win10 64bit and have Cygwin installed that a freind help me some time back to install and to run a script file.  Thats the extent of my Cygwin/Linux knowledge.  I've installed the Cypress software and it appears to see my CY8CKIT-059.
I open your fluxengine-master.zip file and copied/extracted the fluxengine-master folder the folder where I use Cygwin from.
I your Building the Harware instructions you say "When you’re ready, open the FluxEngine.cydsn/FluxEngine.cywrk workspace, pick ‘Program’ from the menu," but there is NO FluxEngine.cywrk file, only FluxEngine.cydwr and FluxEngine.cyprj and a few other files.  If I type bash FluxEngine.cyprj all I get is a very long stream of "no such directory".  Is bash the right way to start these ?  What am I be doing wrong ?  Any help appreciated :)
I'm hoping to image a heap of 5.25" and 3.5" Microbee floppies.  These are 40 or 80 track CP/M disks and I have imaged some of them successfully with an old PC and Dave Dunfields Imagedisk program.
Thanks, Alan

**Comment by davidgiven** [comment]
Sorry, that's a typo. Open `FluxEngine.cyprj` in the Psoc Creator app and it should load the project. You should be able to build and program the board from there.

The good news is that I've discovered a way to provide precompiled binaries, so with luck it shouldn't be necessary to build the firmware yourself soon.

**Comment by ChickenMan540** [comment]
Thanks David, will wait for the Binaries then :)

**Comment by davidgiven** [comment]
I mean 'soon' as in O(small number of weeks)! You probably want to build it yourself right now.

**Comment by ChickenMan540** [comment]
Okay, have opened FluxEngine.cyprj in the Psoc Creator and hit Program.  Things flash in the window (to quick to read) and ends with  sdb.M0052:Unable to find component "cy_boot_v5_80".

**Comment by davidgiven** [comment]
Urgh, it needs an update (Cypress keep changing things). Right click on the project, pick 'Update project', and follow the obvious prompts. Once done it should build.

**Comment by ChickenMan540** [comment]
Thanks, did the upgrade and saw cy_boot one of them updated and 2 others.  Ran Program again and a whole lot of stuff on screen flashed by and ended with a warning
sta.M0019:Warning-1366: Setup time violation found in a path from clock ( CyBUS_CLK ) to clock ( CyBUS_CLK ).

Is that okay ?  The blue led on the main board is now solid and not flashing.  Is that as expected?  Thanks, Alan

**Comment by davidgiven** [comment]
You should be good to go. Try the `fe-testbulktransport` step described in step 7 here: http://cowlark.com/fluxengine/doc/using.html (You don't need to have the board connected to a floppy disk drive.) It'll connect to the board and measure your USB bandwidth.

**Comment by ChickenMan540** [comment]
Okay, that tute says 7. Do .obj/fe-testbulktransport from the shell. It’ll measure your USB bandwidth.  I dont know what "the shell" is or how to run/access it.  What program do I run ?  Thanks Alan

**Comment by davidgiven** [comment]
You need to start the Cygwin console, build the client software there with the `make` command and then run the various client tools from the command line. I'm afraid that knowledge of the command line is required here; there's no GUI and everything's fairly rough edged.

**Comment by ChickenMan540** [comment]
Okay, copy folder in fluxengine-mater.zip to my cygwin work area.  Typed in make and get this error

ChickenMan@CHICKENMAN ~/fluxengine
$ make
meson .obj
process_begin: CreateProcess(NULL, meson .obj, ...) failed.
make (e=2): The system cannot find the file specified.
make: *** [.obj/build.ninja] Error 2

Reading under Building the Client, it says "You’ll need the sqlite3, meson and ninja packages "  How do I install those?
Thanks, Alan

**Comment by ChickenMan540** [comment]
Had a readup on Cygwin and to install those packages, its done via installation of Cygwin.  Anyway upgraded cygwin to the latest and selected the 3 packages to be installed also.  Then ran make again and get this error

ChickenMan@CHICKENMAN ~/fluxengine
$ make
meson .obj
process_begin: CreateProcess(NULL, python3 D:\cygwin\bin\meson .obj, ...) failed.
make (e=2): The system cannot find the file specified.
make: *** [.obj/build.ninja] Error 2

Python3 is also installed.  What do you think the problem could be ?  Thanks, Alan

**Comment by davidgiven** [comment]
Did you install the cygwin version of python3 or a third-party one? Also, one thing the docs should make clearer is that on Cygwin you need the -devel versions of the packages (they contain the files needed to build against them, rather than just run precompiled binaries).

**Comment by ChickenMan540** [comment]
Python3 installed automatically when Cygwin was installed.  I just installed the 32bit version from https://www.cygwin.com/  I cant seam to find anything about a -devel version, where can I get that from?
Thanks, Alan

**Comment by davidgiven** [comment]
Running the cygwin setup program will give you a list of available packages you can install (or remove, or upgrade). The UI is terrible but it does sort of work. See https://cygwin.com/cygwin-ug-net/setup-net.html#internet-setup for the documentation.

**Comment by ChickenMan540** [comment]
Thanks David, thats exactly what I did and you right the UI is terrible.  I just tried un-installing and re-installing the 4 in question, sqlite3, meson, ninja and python3. Unfortunately I get the same error.  I have these version installed
cygwin 3.0.7-1
meson  0.49.2-1
ninja 1.9.0-1
sqlite3 3.28.0-1

**Comment by davidgiven** [comment]
Remember you need the sqlite3-devel package (and, possibly, the libusb-devel package?).

Can you run python3 from the command line?

**Comment by ChickenMan540** [comment]
I installed everything starting with sqlite3 and also libusb-devel packages.  Same error.  Python3 runs fine from the command line.
If it type in meson .obj, I get

ChickenMan@CHICKENMAN ~/fluxengine
$ meson .obj
The Meson build system
Version: 0.49.2
Source dir: /home/ChickenMan/fluxengine
Build dir: /home/ChickenMan/fluxengine/.obj
Build type: native build
Project name: fluxclient
Project version: undefined

meson.build:1:0: ERROR:  Compiler gcc can not compile programs.

A full log can be found at /home/ChickenMan/fluxengine/.obj/meson-logs/meson-log.txt

and the meson-log.txt file reads

Build started at 2019-06-27T09:12:09.836446
Main binary: /usr/bin/python3
Python system: CYGWIN_NT-10.0-WOW
The Meson build system
Version: 0.49.2
Source dir: /home/ChickenMan/fluxengine
Build dir: /home/ChickenMan/fluxengine/.obj
Build type: native build
Project name: fluxclient
Project version: undefined
Sanity testing C compiler: gcc
Is cross compiler: False.
Sanity check compiler command line: gcc /home/ChickenMan/fluxengine/.obj/meson-private/sanitycheckc.c -o /home/ChickenMan/fluxengine/.obj/meson-private/sanitycheckc.exe
Sanity check compile stdout:

-----
Sanity check compile stderr:
gcc.exe: error: /home/ChickenMan/fluxengine/.obj/meson-private/sanitycheckc.c: No such file or directory
gcc.exe: fatal error: no input files
compilation terminated.

-----

meson.build:1:0: ERROR:  Compiler gcc can not compile programs.

Thanks, Alan

**Comment by davidgiven** [comment]
Well, it's an improvement... at least meson's running now.

Try deleting the fluxengine/.obj directory and trying again; also, meson is saying that it's created a  /home/ChickenMan/fluxengine/.obj/meson-private/sanitycheckc.c file, but gcc says it's not there, so it's worth checking that meson is properly creating it.

**Comment by ChickenMan540** [comment]
The .obj folder did contain a sanitycheckc.c file whos contents was
int main(int argc, char **argv) { int class=0; return class; }

I deleted the .obj folder and ran MAKE again, it produced an .obj folder but this time its empty :(
Thanks, Alan

**Comment by davidgiven** [comment]
Okay, so this is really weird --- it looks like gcc is failing to see the files which Cygwin has created!

The most plausible explanation for this is if you have a non-Cygwin gcc which for some reason is being used in preference to the Cygwin ones. This was why I asked about Python3 earlier. But Cygwin shouldn't actually let that happen.

Can you compile that sanitycheck.c program manually from the command line? `gcc /home/ChickenMan/fluxengine/.obj/meson-private/sanitycheckc.c`? If that fails too, try `which gcc` and see which one it's finding.

**Comment by ChickenMan540** [comment]
Okay, I ran the command line above and got
ChickenMan@CHICKENMAN ~/fluxengine
$ gcc /home/ChickenMan/fluxengine/.obj/meson-private/sanitycheckc.c
gcc.exe: error: /home/ChickenMan/fluxengine/.obj/meson-private/sanitycheckc.c: No such file or directory
gcc.exe: fatal error: no input files
compilation terminated.

I then included d:/cygwin in the command line and got no error

ChickenMan@CHICKENMAN ~/fluxengine
$ gcc D:/cygwin/home/ChickenMan/fluxengine/.obj/meson-private/sanitycheckc.c

ChickenMan@CHICKENMAN ~/fluxengine

I then did a which gcc

$ which gcc
/cygdrive/d/Strawberry/c/bin/gcc

ChickenMan@CHICKENMAN ~/fluxengine

I just copied the FluxEngine folder to my Cygwin normal work area, do I have to give it any special Read/Write permissions or anything like that ?
Thanks, Alan

**Comment by davidgiven** [comment]
Right! What's happening is that it's picking up the non-cygwin gcc (in D:/Strawberry), rather than the Cygwin version (in D:/cygwin/usr/bin), which doesn't have the modifications needed to work under cygwin. If you do `echo $PATH` you will probably see /cygdrive/d/Strawberry/c/bin appear before /usr/bin.

The quick fix is straightforward: rename D:/Strawberry to D:/Not.Strawberry, so that Cygwin can't find it. The longer term fix is to adjust your Cygwin path so that /usr/bin appears first.

Cygwin is a mess.

**Comment by ChickenMan540** [comment]
I renamed my D:|Strawberry folder and then which gcc cant find any at all.  So I re-installed cygwin adding gcc so a which gcc produces
$ which gcc
/usr/bin/gcc

So thats sorted. I then deleted the .obj folder and ran make again, unfortunately I still get the same error and .obj is empty :(

$ make
meson .obj
process_begin: CreateProcess(NULL, python3 D:\cygwin\bin\meson .obj, ...) failed.
make (e=2): The system cannot find the file specified.
make: *** [.obj/build.ninja] Error 2

Thanks, Alan

**Comment by ChickenMan540** [comment]
I did a which xxx with xxx being the extra programs needed to be installed.  All good except for libusb which cant be found.  Re-installed cygwin but it says its already install, so I installed everything that started with libusb.  Still cant be found :(

**Comment by davidgiven** [comment]
Sorry for the delay --- I've actually done a big update and changed the way the build system works to be a lot simpler. You don't need meson any more, which should help, and the definitive list of Cygwin packages you need is:

libsqlite3-devel, ninja, zlib-devel, libusb1.0-devel

(I know this because I now have it building on the continuous integration servers, and here's the log for proof: https://ci.appveyor.com/project/davidgiven/fluxengine)

I'm also going to attempt to generate precompiled binaries which can just be installed somewhere and which don't require Cygwin, because you really shouldn't have to go through this; but that'll take a little while to make work.

**Comment by ChickenMan540** [comment]
Thanks for that, I will grab the new update, add in a couple of the new packages for Cygwin and give it a go this afternoon when home.
Pre-compiled binaries is the way to go in my opinion for your project to really "take off". The average Windows user can run the binaries but trying to generate them and with Cygwin, is a MAJOR learning curve filled with lots of problems, as seen in my experience.  I look forward to their arrival :)
Thanks, Alan

**Comment by ChickenMan540** [comment]
Woo hoo, the new script compiled with no issues :) 

ChickenMan@CHICKENMAN ~/fluxengine
$ make
MKNINJA .obj/build.ninja
[108/108] STRIP fluxengine-debug

Plugged the hardware in via the small usb and tried, as you suggested, "Try the fe-testbulktransport step described in step 7 " with  **fluxengine testbulktransport** from the shell and get this error

ChickenMan@CHICKENMAN ~/fluxengine
$ fluxengine testbulktransport
bash: fluxengine: command not found

But there is a fluxengine.exe file in the fluxengine folder. Its 578,574 bytes long.

I think I might just wait for your installable binaries :)

Thanks, Alan

**Comment by lanroth** [comment]
You need to type: ./fluxengine
The current working directory isn't on the path in Unix.
https://superuser.com/questions/156582/why-is-not-in-the-path-by-default

**Comment by ChickenMan540** [comment]
Arrr... thanks, yep that worked and I get 749 kB/s :)   Now I need to set up a drive and image an IBM disk to start with.
Thanks, Alan

**Comment by ChickenMan540** [comment]
Reading the docs, 749 kB/s might be a bit slow, so tried other USB ports and they all give "cant find FlugEngine (is it plugged in?)"   Yes its plugged in and Windows has it registered in Device Manager.  Is this yet another Cygwin/Linux quirk ?
Thanks, Alan

**Comment by lanroth** [comment]
I've had this a lot working with microcontrollers and Windows. Try rebooting if you haven't - that almost always fixes it for me.

**Comment by davidgiven** [comment]
Good news! There's now a precompiled version of the new FluxEngine client for Windows downloadable from the GitHub releases page at https://github.com/davidgiven/fluxengine/releases/latest. This will (should) be automatically updated whenever things change.

It hasn't had a lot of testing.

**Comment by ChickenMan540** [comment]
Fantastic, a bit tied for a couple of days but will test out asap, many thanks :)  Alan

**Comment by ChickenMan540** [comment]
Finally gave it a run,  
fluxengine testbulktransport still gives me 749k/s  
fluxengine rpm gives me 300 rpm
fluxengine read ibm  read a blank 1.44mb ibm formated disk in a Sony 920 floppy with no errors

1.14 ................................................................................
1.15 ................................................................................
1.16 ................................................................................
1.17 ................................................................................
Good sectors: 2880/2880 (100%)
Missing sectors: 0/2880 (0%)
Bad sectors: 0/2880 (0%)
80 tracks, 2 heads, 18 sectors, 512 bytes per sector, 1440 kB total

G:\FluxEngine>fluxengine rpm
Rotational period is 200 ms (300 rpm)

G:\FluxEngine>

I tried it on a MS Mouse drivers disk and if read to an ibm.img file (though 2 bad sectors) and looking at it with a HexEditor, it all looks great.
Also tried an MS-DOS 6.22 Disk 1 and got a clean error free image again.  So looking good for IBM disks.

Now to read some Microbee CP/M floppies.  The main 3 styles of disks are 
1. 80 track double sided 10 sector/tr 512 b/s, 
2. 80 track Single sided 10 sector/tr 512 b/s  
3. 40 track double sided 10 sectors/tr 512 b/s

What do I need to type in to image these to a Flux file ?
Thanks, Alan

**Comment by davidgiven** [comment]
Fantastic.

So, for the Microbee formats: apparently they're standard IBM scheme, so the first thing to try is `fluxengine read ibm` on them. For the single sided disk you'll want `-s :s=0` and for the 40 track disk you'll want `-s t=0-79x2` or `-s t=1-79x2` (one of these will give a better read than the other). Almost everything else should be autodetected. One thing which can't be autodetected is whether the sector numbers start with 0 or 1, so if you see a disk with nine sectors per track try `--sector-id-base=0`.

However, bear in mind that CP/M can also applies sector skew in software, so that logical sector 1 isn't necessarily stored in physical sector 1. This information isn't stored on the disk anywhere and FluxEngine can't help with that (and isn't intended to). It's the job of whatever tool you use to extract files from the disk image to handle that --- I use cpmtools for this, and it does actually support Microbee disks.

Some Microbee disks apparently used different sector numbering on different tracks (from https://github.com/keirf/FlashFloppy/issues/105). This will need custom code to handle...

**Comment by ChickenMan540** [comment]
I tried a couple of standard Microbee single sided 80 tr 3.5" disks (SS80) with 
**fluxengine read ibm -s :s=0** 
and I can load the images into ubee512 emulator with all files seen and can run and can be extracted :) .  Fantastic.  I'm sure the 40 tr 5.25" disks will be the same, will test later though. 

Your right with some disks (the DS80) have first 5 tracks sectors 1 to 10 then 6 to 79 sectors start at 21.  Microbee also produced many copy protected disks, they are 40 tr single sided with sectors 1 to 10 except track 5 which has sectors 0 to 9.  I tried one of these but bytes/sector was sen as 1024 bytes and not 512 so final file is 400kb when it should be 200kb.

39.0: 210 ms in 76864 bytes
       60 records, 30 sectors; 2.06us clock;
       Failed to read sector 3 (bad checksum);
       giving up
       logical track 39.0; 5120 bytes decoded.
H.SS Tracks --->
0. 0 ........................................
0. 1 ........................................
0. 2 ........................................
0. 3 .......................BBBB.......BBBBBB
0. 4 ........................................
0. 5 .....XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
0. 6 .....XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
0. 7 .....XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
0. 8 .....XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
0. 9 .....XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
Good sectors: 215/400 (53%)
Missing sectors: 175/400 (43%)
Bad sectors: 10/400 (2%)
40 tracks, 1 heads, 10 sectors, 1024 bytes per sector, 400 kB total

G:\FluxEngine>

CPMTOOLS works with Microbee DS40 disks (double sided 40 tr 5.25") only and does not with SS80 or DS80 disks.  But there is a patched cpmtools that can access them all.

So some great success, will try a few more tomorrow :)  Thanks, Alan

**Comment by davidgiven** [comment]
The algorithm for the final image used is to pick the maximum sector size and sector count on the disk. So that DS80 disk looks like there are five tracks of 512 byte sectors and then the rest is 1024 byte sectors. The output image will have each 512 byte sector padded with zeroes to make them 1024 bytes long, while tracks 6 and above will be padded with zeroes to make them ten sectors long.

This is unfortunate but it's the best I could do at the time. Handling this better is on my list of things to do.

**Comment by ChickenMan540** [comment]
Okay, no problems, I'm sure its operation will improve with time :)

One question, the fluxengine read ibm produces an ibm.img file, is there any way to allow you to name the output file.  If I have a dozen disks to image, I have to image one disk, then rename the ibm.bin file, before imaging the next, otherwise it gets over written? 
Thanks, Alan

**Comment by ChickenMan540** [comment]
Just tried imaging a Microbee 5.25" 40 double sided disk.  I've tried both 360kb 40 tr and a 1.2mb 80 tr drives (double stepping the latter) and get same results.  It reads side 0 okay but fails side 1 with

2.0: 174 ms in 79360 bytes
       20 records, 10 sectors; 1.75us clock;
       logical track 1.0; 5120 bytes decoded.
  2.1: 174 ms in 81728 bytes
       20 records, 10 sectors; 1.75us clock;
       logical track 1.0;
       multiple conflicting copies of sector 0 seen;
       multiple conflicting copies of sector 1 seen;
       multiple conflicting copies of sector 2 seen;
       multiple conflicting copies of sector 3 seen;
       multiple conflicting copies of sector 4 seen;
       multiple conflicting copies of sector 5 seen;
       multiple conflicting copies of sector 6 seen;
       multiple conflicting copies of sector 7 seen;
       multiple conflicting copies of sector 8 seen;
       multiple conflicting copies of sector 9 seen; 5120 bytes decoded.

I suspect the problem to be that both side 0 and side 1 are designated as side 0.  Thats another issue for your to-do list :)
Thanks, Alan

**Comment by davidgiven** [comment]
Argh. Yes, looks like the Microbee used the same head number on both sides of the disk. I haven't seen an IBM scheme disk that did that... it's either a very easy fix or a moderately complex one, and because it's all the generic IBM decoder it'll affect lots of formats.

**Comment by ChickenMan540** [comment]
Hi David,  another question please if I may, can FluxEngine read Kryoflux RAW Stream files to an IMG ?  I tried  
G:\FluxEngine>fluxengine read ibm -s track00.0.raw:t=0-79
Reading from: track00.0.raw:d=0:s=0-1:t=0-79
Error: unrecognised flux filename extension

With track00.0.raw the first of 160 files.  
Thanks, Alan

**Comment by davidgiven** [comment]
Yes --- but you need to point it at the _directory_ containing the track files, not the track file itself. There are also some rough edges of you refer to a track which isn't present, so be careful with the `t=` clause.

**Comment by ChickenMan540** [comment]
Thanks, I put all in a folder called **vista_mitsu_055** in my FluxEngine folder and tried
G:\FluxEngine>**fluxengine read ibm -s vista_mitsu_055**
Reading from: vista_mitsu_055:d=0:s=0-1:t=0-79
Error: unrecognised flux filename extension

If I used the full path to the folder I get

G:\FluxEngine>**fluxengine read ibm -s g:\FluxEngine\vista_mitsu_055**
Error: invalid data modifier syntax '\FluxEngine\vista_mitsu_055'

Oh, and these are Exidy Sorcerer disks with Vista controller files from a friend.
Thanks, Alan

**Comment by ChickenMan540** [comment]
Okay, eventually got things to happen with
G:\FluxEngine>fluxengine read ibm -s /fluxengine/vista_mitsu_055/
Reading from: /fluxengine/vista_mitsu_055/:d=0:s=0-1:t=0-79
  0.0: 1208 ms in 470781 bytes
       0 bytes decoded.
  0.1: 1208 ms in 446952 bytes
       0 bytes decoded.
  1.0: 1204 ms in 456297 bytes
       0 bytes decoded.
  1.1: 1209 ms in 432713 bytes
       0 bytes decoded.
.
.
.
 79.0: 1199 ms in 438232 bytes
       0 bytes decoded.
 79.1: 1214 ms in 226555 bytes
       0 bytes decoded.
H.SS Tracks --->
No sectors in output; skipping analysis
0 tracks, 0 heads, 0 sectors, 0 bytes per sector, 0 kB total

but no decoding and a 0 byte ibm.img.
I think these are from Hard Sector disks so thats probably the problem.
Thanks, Alan

**Comment by davidgiven** [comment]
Yep, looks like a custom format --- that'll need it's own decoder. Unless it's the same as one of the existing formats!

I see this: https://forum.kryoflux.com/viewtopic.php?t=1354

If you can share the files with me I can have a look, but it'll be a while before I can get to it. OTOH you're welcome to have a go yourself! Hard sector formats are pretty easy to decode; take a look at one of the existing decoders...

**Comment by davidgiven** [comment]
BTW, you can combine --write-flux with Kryoflux streams.

```
$ fluxengine read ibm --just-read --write-flux=out.flux -s some/kryoflux/stream/
```

...will convert the Kryoflux stream to a flux file, which is both smaller and easier to send.
